FEXCore/Source/Interface/Core/CodeCache.cpp

```diff
@@ -225,7 +231,80 @@ void CodeCache::LoadData(Core::InternalThreadState& Thread, std::byte* MappedCac
 }
 
 bool CodeCache::SaveData(Core::InternalThreadState& Thread, int fd, const ExecutableFileSectionInfo& SourceBinary, uint64_t SerializedBaseAddress) {
-  // TODO
+  auto CodeBuffer = CTX.GetLatest();
+  auto& LookupCache = *Thread.LookupCache->Shared;
+
+  auto Relocations = Thread.CPUBackend->TakeRelocations(SourceBinary.FileStartVA);
+
+  // Write file header
+  struct Header {
+    char Magic[4] = {'F', 'X', 'C', 'C'};
+    uint32_t FormatVersion = 1;
+    char FEXVersion[8] = {};
+    uint32_t NumBlocks;
+    uint32_t NumCodePages;
+    uint32_t CodeBufferSize;
+    uint32_t NumRelocations;
+    uint64_t SerializedBaseAddress;
+    // TODO: Consider including information from LookupCache.BlockLinks
+  } header;
+  memcpy(&header.FEXVersion[0], GIT_SHORT_HASH, strlen(GIT_SHORT_HASH));
+  header.NumBlocks = LookupCache.BlockList.size();
+  header.NumCodePages = LookupCache.CodePages.size();
+  header.CodeBufferSize = CTX.LatestOffset;
+  header.NumRelocations = Relocations.size();
+  header.SerializedBaseAddress = SerializedBaseAddress;
+  ::write(fd, &header, sizeof(header));
+
+  // Dump guest<->host block mappings
+  {
+    fextl::vector<decltype(LookupCache.BlockList)::value_type> BlockList {LookupCache.BlockList.begin(), LookupCache.BlockList.end()};
+    for (auto [Guest, Host] : LookupCache.BlockList) {
+      Guest -= SourceBinary.FileStartVA;
+      ::write(fd, &Guest, sizeof(Guest));
+      Host.HostCode -= reinterpret_cast<uintptr_t>(CodeBuffer->Ptr);
+      ::write(fd, &Host.HostCode, sizeof(Host.HostCode));
+      uint64_t NumCodePages = Host.CodePages.size();
+      ::write(fd, &NumCodePages, sizeof(NumCodePages));
+      LOGMAN_THROW_A_FMT(std::ranges::is_sorted(Host.CodePages), "Code pages aren't sorted");
+      for (auto CodePage : Host.CodePages) {
+        CodePage -= SourceBinary.FileStartVA;
+        ::write(fd, &CodePage, sizeof(CodePage));
+      }
+    }
+  }
+
+  // Dump relocations
+  ::write(fd, Relocations.data(), Relocations.size() * sizeof(Relocations[0]));
+
+  // Pad to next page in file so that the CodeBuffer can be mmap'ed into process on load
+  // TODO: lseek not currently available with our custom CRT
+#ifndef _WIN32
```

Comment:
You can just add this within the PR (https://github.com/bylaws/FEX/blame/arm64eccc/Source/Windows/Common/CRT/IO.cpp#L203) we already have _lseek so just adding alias.  Also note this code doesn't even need to work with the FEX CRT, just needs to link successfully. The windows offline compiler can depend on the standard CRT.